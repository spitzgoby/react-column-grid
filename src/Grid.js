import classNames from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';

import './Grid.scss';

const numCols = 12;
const defaultWidths = { xs: 12 };
const sizes = ['xs', 'sm', 'md', 'lg', 'xl'];

const Grid = ({
    children,
    className,
    container,
    fullRow = false,
    item,
    offset = {},
    width = { xs: 12 }
}) => {
    const widths = { ...defaultWidths, ...width };
    
    const areValidColumns = (colWidth, colOffset) =>
        !isNaN(parseInt(colWidth, 10)) &&
        !isNaN(parseInt(colOffset, 10)) &&
        colWidth > 0 && 
        colOffset + colWidth <= numCols;
   
    const calculateColSizes = () => 
        sizes.reduce((acc, size) => {
            const colWidth = widths[size];
            const colOffset = offset[size];
            
            return (areValidColumns(colWidth, colOffset)) 
                ? acc.concat([`swa-react-grid--item_${size}_${colOffset}_${colWidth}`]) 
                : acc;
        }, []);
    

    const getClass = () => {
        return classNames(
            className, 
            {
                'swa-react-grid': true,
                'swa-react-grid--container': container,
                'swa-react-grid--item': item,
            }, 
            calculateColSizes()
        );
    };

    const getAdjustedLayoutProps = (size, child, columns) => {
        let adjustedColumn;
        let adjustedOffset;
        const { width, offset } = child.props;
        const sizeOffset = offset?.[size] || 0;
        const sizeWidth = width?.[size];
        const suggestedOffset = columns[size] + sizeOffset;

        if (areValidColumns(suggestedOffset, sizeWidth)) {
            adjustedOffset = suggestedOffset;
            adjustedColumn = (suggestedOffset + sizeWidth) % numCols;
        } else {
            adjustedOffset = sizeOffset;
            adjustedColumn = columns[size];
        }

        return {
            adjustedOffset,
            adjustedColumn
        };
    };

    const renderChildren = () => {
        let currentColumns = { xs: 0, sm: 0, md: 0, lg: 0, xl: 0 };

        return children?.map((child, index) => {
            let renderedChild = child;

            if (child.type.name === 'Grid' && child.props.item) {
                const renderedChildOffset = sizes.reduce((acc, size) => {
                    const { 
                        adjustedColumn, 
                        adjustedOffset 
                    } = getAdjustedLayoutProps(size, child, currentColumns); 

                    currentColumns[size] = adjustedColumn;
                    acc[size] = adjustedOffset;

                    return acc;
                }, {});
                const renderedChildProps = {
                    ...child.props,
                    key: index,
                    offset: renderedChildOffset
                };
                
                renderedChild = <Grid { ...renderedChildProps } />;
            }

            return renderedChild;
        });
    };

    return (
        <div className={getClass()}>{container ? renderChildren() : children}</div>
    );
};

Grid.propTypes = {
    container: PropTypes.bool,
    fullRow: PropTypes.bool,
    item: PropTypes.bool,
    offset: PropTypes.shape({
        xs: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
        sm: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
        md: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
        lg: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
        xl: PropTypes.oneOfType([PropTypes.number, PropTypes.string])
    }),
    width: PropTypes.shape({
        xs: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
        sm: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
        md: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
        lg: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
        xl: PropTypes.oneOfType([PropTypes.number, PropTypes.string])
    })
};

export default Grid;