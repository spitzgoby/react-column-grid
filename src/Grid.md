# Grid

## container

**Type**: `PropTypes.bool`  
**Default**: `false`  
**Description**: Determines whether the grid component is a container. All grids
should have a container element at the root. The container element will have its
`display` attribute set to `grid`.

**Usage**

```Javascript
<Grid container />
```

---

## columns

**Type**: `PropTypes.oneOfType([PropTypes.number, PropTypes.string])`
**Default**: `12`
**Description**: Sets the number of columns to use when laying out the grid items.
This value is ignored if the `container` prop is `false`.
**Usage**

```Javascript
<Grid container columns={6}>
    <Grid item width={{ xs: 3 }} />
    <Grid item width={{ xs: 3 }} />
    <Grid item width={{ xs: 3 }} /> {/* this item will be moved onto a new row */}
</Grid>
```

---

## item

**Type**: `PropTypes.bool`  
**Default**: `true`  
**Description**: Determines whether the grid component is an item within a container.
An item will have `grid-column` properties assigned based on the width and offset
prop values.  
**Usage**

```Javascript
<Grid item />
```

---

## width

**Type**:

```Javascript
PropTypes.shape({
    xs: PropTypes.oneOf([PropTypes.number, PropTypes.string]),
    sm: PropTypes.oneOf([PropTypes.number, PropTypes.string]),
    md: PropTypes.oneOf([PropTypes.number, PropTypes.string]),
    lg: PropTypes.oneOf([PropTypes.number, PropTypes.string]),
    xl: PropTypes.oneOf([PropTypes.number, PropTypes.string])
})
```

**Default**:

```Javascript
{
    xs: 12,
    sm: 0,
    md: 0,
    lg: 0,
    xl: 0
}
```

**Description**: Sets the number of columns a grid item spans for a given breakpoint.
If the value is not defined at a given break point then the value for the next
smallest breakpoint will be assigned. This prop is ignored if the `item` prop is
`false`. In order to apply the same width at all breakpoints simply provide a single
numeric value.

**Usage**:

```Javascript
// Creates a grid with a predefined width at every breakpoint
<Grid item width={{
    xs: '12',
    sm: '10',
    md: '8',
    lg: 6,
    xl: 4
}} />

// No value is set for sm or xl. The sm width will inherit a width of 12 from
// the xs value, and lg will inherit 8 from the md value.
<Grid item width={{ xs: 12, md: 8, xl: 6 }} />

// Creates an item of width 12 at all breakpoints.
<Grid item width='12' />
```

---

## offset

**Type**:

```Javascript
PropTypes.shape({
    xs: PropTypes.oneOf([PropTypes.number, PropTypes.string]),
    sm: PropTypes.oneOf([PropTypes.number, PropTypes.string]),
    md: PropTypes.oneOf([PropTypes.number, PropTypes.string]),
    lg: PropTypes.oneOf([PropTypes.number, PropTypes.string]),
    xl: PropTypes.oneOf([PropTypes.number, PropTypes.string])
})
```

**Default**:

```Javascript
{
    xs: 0,
    sm: 0,
    md: 0,
    lg: 0,
    xl: 0
}
```

**Description**: Sets the number of columns that the given item should be spaced from
the start of the grid at a given breakpoint. If no value is set at a given breakpoint
then the value from the next smallest breakpoint will be used. This prop is ignored
if the `item` prop is `false`. In order to apply the same offset at all breakpoints
simply provide a single numeric value.

**Usage**:

```Javascript
// Creates an item that is full width on xs and sm screens, 66% width and centered
// on md screens, and 50% and centered on lg or xl screens.
<Grid item  offset={{ md: 4, lg: 3 }} width={{ xs: 12, md: 8, lg: 6 }} />

// Creates an item with an offset of 4 across all breakpoints
<Grid item offset={4} width={8} />
```

---

## clear

**Type**:

```Javascript
PropTypes.shape({
    xs: PropTypes.bool,
    sm: PropTypes.bool,
    md: PropTypes.bool,
    lg: PropTypes.bool,
    xl: PropTypes.bool
})
```

**Default**:

```Javascript
{
    xs: false,
    sm: false,
    md: false,
    lg: false,
    xl: false
}
```

**Description**: Determines whether a given item should take up the remaining space
of he row that it occupies even if its width and offset would not normally span an
entire row. If no value is set at a given breakpoint the value from the next
smallest breakpoint will be used. This value is ignored if the `item` prop is
`false`.

**Usage**:

```Javascript
// Creates a grid that has 2 rows on xs and sm screens, but only a single row on md
// and larger screens.
<Grid container>
    <Grid item
        clear={{ xs: true, md: false }}
        width={{ xs: 6, md: 3 }}
        offset={{ xs: 3 }} />
    <Grid item
        clear={{ xs: true, md: false }}
        width={{ xs: 6, md: 3 }}
        offset={{ xs: 3 }} />
</Grid>
```

---

## hide

**Type**:

```Javascript
PropTypes.shape({
    xs: PropTypes.bool,
    sm: PropTypes.bool,
    md: PropTypes.bool,
    lg: PropTypes.bool,
    xl: PropTypes.bool
})
```

**Default**:

```Javascript
{
    xs: false,
    sm: false,
    md: false,
    lg: false,
    xl: false
}
```

**Description**: Allows grid items or containers to be hidden at specific breakpoints.
If a value is provided at a given breakpoint then it will be applied to all breakpoints
that size or larger until a `false` value is explicitly declared. This item will be
given the `display: none` attribute.

**Usage**:

```Javascript
// Creates an item that only appears at the md break point.
<Grid item hide={{ xs: true, md: false, lg: true }}></Grid>

// Creates an item that is only displayed on xs or md screens
<Grid item hide={{ md: true }}></Grid>
```

---

## gap

**Type**: `PropTypes.oneOf([PropTypes.string, PropTypes.number])`  
**Default**: `'1em'`
**Description**: Determines the size of the grap between grid columns. The default
unit is `em`, but a string can be provided with an arbitrary unit. The component
does not validate the string and will happily apply any arbitrary value as the gap.
This value is ignored if the component is not a `container`.

**Usage**:

```Javascript
// Sets the grid-gap to 20px for the container
<Grid container gap="20px"></Grid>
```
