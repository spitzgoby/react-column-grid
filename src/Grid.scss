.swa-react-grid {
    &--container {
        grid-template-columns: repeat(12, 1fr);
    }
}