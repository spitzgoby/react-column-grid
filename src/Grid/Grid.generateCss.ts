import { BreakpointsDefinition } from "../utils/breakpoints";
import { generateScreenQueries } from "../utils/mediaQueries";
import { sizes } from "../utils/breakpoints";

const containerDisplay = 'display: grid;';
export const generateGridContainerCss = (columns: number) => {
    const containerTemplateColumns = `grid-template-columns: repeat(${columns}, 1fr);`;

    return `.rcg-c{${containerDisplay}${containerTemplateColumns}}`;
}

export const generateGridBreakpointCss = (breakpoints: BreakpointsDefinition, columns: number) => {
    let columnSpanClassDefinition: string[] = sizes.map((size) => {
        let sizeDefintions: string[] = [`.rcg-${size}_h{display: none;}`];        

        for (let col = 1; col <= columns; col++) {
            for (let width = 1; width <= columns; width++) {
                if (col + width <= columns + 1) {
                    sizeDefintions.push(`.rcg-${size}-${col}-${width}{grid-column:${col}/ span ${width};}`) ;
                } else {
                    width = columns;
                }
            }
        }

        return sizeDefintions.join('');
    });

    return generateScreenQueries(breakpoints, columnSpanClassDefinition);
}