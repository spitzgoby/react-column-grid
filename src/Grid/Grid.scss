@import '../style/sizes.scss';

.rcg-c {
  display: grid;
  grid-template-columns: repeat(12, 1fr);
}

@media screen and (min-width: 1535px) {
  .rcg-xl_h {
    display: none;
  }
}

@each $sizeName, $maxWidth in $maxWidthSizes {
  @media screen and (max-width: #{$maxWidth}) {
    .rcg-#{$sizeName}_h {
      display: none;
    }
  }
}

@each $sizeName, $minWidth in $minWidthSizes {
  @media screen and (min-width: #{$minWidth}) {
    @for $col from 1 through 12 {
      @for $width from 1 through 12 {
        @if ($col + $width <= 13) {
          .rcg-#{$sizeName}-#{$col}-#{$width} {
            grid-column: #{$col} / span #{$width};
          }
        }
      }
    }
  }
}