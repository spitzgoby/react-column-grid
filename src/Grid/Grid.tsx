import { 
    getValueOfNumeric, 
    Numeric, 
    numericIsInteger 
} from '../utils/numeric';
import classNames from "classnames";
import {
    BooleanBreakpointValues,
    generateContainerClassName,
    generateSizeClassNames,
    getAdjustedLayoutProps,
    NumericBreakpointValues
} from "./Grid.layout";
import PropTypes, { ReactNodeLike } from "prop-types";
import React, { useContext } from "react";
import { GridProvider, GridContext } from '../GridProvider';
import { addMissingSizes, DEFAULT_SCREEN_WIDTHS, ScreenWidths, sizes } from "../utils/breakpoints";

import { DEFAULT_GAP } from "../utils/gap";
import { DEFAULT_COLUMNS, INITIAL_ID } from '../GridProvider/GridContext';

const defaultClear = false;
const defaultHide = false;
const defaultOffset = 0;
const defaultWidth = 12;
const booleanShortHandProps = ["clear"];
const numericShortHandProps = ["offset", "width"];

const useShorthandSyntax = (propName: string, prop: Numeric | boolean): boolean => {
    let result = false;    

    if (typeof prop === 'boolean') { 
        result = booleanShortHandProps.includes(propName);
    } else if (numericShortHandProps.includes(propName)) {
        result = numericIsInteger(prop);
    }

    return result;
}

type Props = {
    breakpoints?: ScreenWidths,
    children: ReactNodeLike,
    className?: string,
    clear?: boolean | BooleanBreakpointValues,
    columns?: number,
    container?: boolean,
    gap?: Numeric,
    hide?: BooleanBreakpointValues,
    item?: boolean,
    offset?: NumericBreakpointValues | Numeric,
    width?: NumericBreakpointValues | Numeric,
}

const Grid: React.FC<Props> = (props) => {
    const {
        breakpoints: propBreakpoints,
        children,
        className,
        columns: propColumns,
        container,
        gap: propGap,
        hide = {},
        item,
        offset = {},
        width = { xs: 12 },
    } = props;
    const { 
        breakpoints: contextBreakpoints, 
        columns: contextColumns, 
        gap: contextGap,
        id 
    } = useContext(GridContext);
    const breakpoints = propBreakpoints || contextBreakpoints || DEFAULT_SCREEN_WIDTHS;
    const columns = propColumns || contextColumns || DEFAULT_COLUMNS;
    const gap = propGap || contextGap || DEFAULT_GAP;    

    // Fill out incomplete layout props
    const adjustedHide = addMissingSizes(
        "hide",
        hide,
        defaultHide,
        useShorthandSyntax
    );
    const adjustedOffset = addMissingSizes(
        "offset",
        offset,
        defaultOffset,
        useShorthandSyntax
    );
    const adjustedWidth = addMissingSizes(
        "width",
        width,
        defaultWidth,
        useShorthandSyntax
    );

    const hasPropsThatOverrideContext = () => 
        breakpoints !== contextBreakpoints
        || columns !== contextColumns
        || gap !== contextGap;
    const shouldRenderWithProvider = () => container && (id === INITIAL_ID || hasPropsThatOverrideContext());

    // Generate classes based on width, offset and sizes
    const sizeClassNames = generateSizeClassNames({
        hide: adjustedHide,
        item,
        offset: adjustedOffset,
        width: adjustedWidth,
    }, id);

    const getClass = () =>
        classNames({ [generateContainerClassName(id)]: container }, sizeClassNames, className);

    const renderChildren = () => {
        let currentColumns = { xs: 0, sm: 0, md: 0, lg: 0, xl: 0 };
        const childrenToRender = Array.isArray(children)
            ? children
            : [children];

        return childrenToRender.map((child, index) => {
            let renderedChild = child;

            if (child?.type?.name === Grid.name) {
                let renderedChildProps = { 
                    ...child.props,
                    key: index
                };

                if (child?.props?.item) {
                    const renderedChildOffset = sizes.reduce((acc: NumericBreakpointValues, size) => {
                        const completedClear = addMissingSizes(
                            "clear",
                            child.props.clear,
                            defaultClear,
                            useShorthandSyntax
                        );
                        const completedHide = addMissingSizes(
                            "hide",
                            child.props.hide,
                            defaultHide,
                            useShorthandSyntax
                        );
                        const completedOffset = addMissingSizes(
                            "offset",
                            child.props.offset,
                            defaultOffset,
                            useShorthandSyntax
                        );
                        const completedWidth = addMissingSizes(
                            "width",
                            child.props.width,
                            defaultWidth,
                            useShorthandSyntax
                        );
                        const { adjustedColumn, adjustedOffset } =
                            getAdjustedLayoutProps(
                                size,
                                completedOffset,
                                completedWidth,
                                completedHide,
                                completedClear,
                                currentColumns,
                                columns
                            );

                        currentColumns[size] = getValueOfNumeric(adjustedColumn);
                        acc[size] = getValueOfNumeric(adjustedOffset);

                        return acc;
                    }, {});

                    renderedChildProps = {
                        ...renderedChildProps,
                        offset: renderedChildOffset,
                    };
                } 

                renderedChild = <Grid {...renderedChildProps} />;
            }

            return renderedChild;
        });
    };

    const renderWithoutGridProvider = () => (
        <div className={getClass()}>
            <>{container ? renderChildren() : children}</>
        </div>
    );

    const renderWithGridProvider = () => (
        <GridProvider breakpoints={breakpoints} columns={columns} gap={gap}>
            <Grid {...props} />
        </GridProvider>
    )

    return (
        shouldRenderWithProvider() ? renderWithGridProvider() : renderWithoutGridProvider()
    );
};

Grid.propTypes = {
    breakpoints: PropTypes.arrayOf(PropTypes.number),
    children: PropTypes.node,
    clear: PropTypes.oneOfType([
        PropTypes.bool, 
        PropTypes.shape({
            xs: PropTypes.bool,
            sm: PropTypes.bool,
            md: PropTypes.bool,
            lg: PropTypes.bool,
            xl: PropTypes.bool,
        })
    ]),
    className: PropTypes.string,
    columns: PropTypes.number,
    container: PropTypes.bool,
    gap: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    hide: PropTypes.shape({
        xs: PropTypes.bool,
        sm: PropTypes.bool,
        md: PropTypes.bool,
        lg: PropTypes.bool,
        xl: PropTypes.bool,
    }),
    item: PropTypes.bool,
    offset: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.string,
        PropTypes.shape({
            xs: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
            sm: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
            md: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
            lg: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
            xl: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
        }),
    ]),
    width: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.string,
        PropTypes.shape({
            xs: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
            sm: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
            md: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
            lg: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
            xl: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
        }),
    ]),
};

export default Grid;
