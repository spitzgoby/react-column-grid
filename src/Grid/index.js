import GridComponent from "./Grid";

export const Grid = GridComponent;
export default GridComponent;
