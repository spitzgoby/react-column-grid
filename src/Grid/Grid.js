import { addMissingSizes, sizes } from "../utils/breakpoints";
import classNames from "classnames";
import {
    generateSizeClassNames,
    getAdjustedLayoutProps,
    getGap,
} from "./Grid.layout";
import PropTypes from "prop-types";
import React from "react";
import ThemeContext from "../ThemeContext";

import "./Grid.scss";

const defaultClear = false;
const defaultHide = false;
const defaultOffset = 0;
const defaultWidth = 12;
const shortHandProps = ["offset", "width"];

const isInteger = (value) => !isNaN(parseInt(value, 10));
const useShorthandSyntax = (propName, prop) =>
    shortHandProps.includes(propName) && isInteger(prop);

const Grid = ({
    children,
    className,
    container,
    gap: propGap = null,
    hide = {},
    item,
    offset = {},
    width = { xs: 12 },
}) => {
    // Determine breakpoints and update state
    // State is used to ensure that the context provider does not trigger
    // unnecessary renders
    const { gap } = React.useContext(ThemeContext);

    const shouldUsePropGap = container && propGap;
    const calculatedGap = shouldUsePropGap ? propGap : gap;
    const [adjustedGap, setAdjustedGap] = React.useState(calculatedGap);

    if (shouldUsePropGap && adjustedGap !== propGap) {
        setAdjustedGap(calculatedGap);
    }

    // Fill out incomplete layout props
    const adjustedHide = addMissingSizes(
        "hide",
        hide,
        defaultHide,
        useShorthandSyntax
    );
    const adjustedOffset = addMissingSizes(
        "offset",
        offset,
        defaultOffset,
        useShorthandSyntax
    );
    const adjustedWidth = addMissingSizes(
        "width",
        width,
        defaultWidth,
        useShorthandSyntax
    );

    // Generate classes based on width, offset and sizes
    const sizeClassNames = generateSizeClassNames({
        hide: adjustedHide,
        item,
        offset: adjustedOffset,
        width: adjustedWidth,
    });

    const getThemeValues = () => ({ gap: adjustedGap });

    const getClass = () =>
        classNames({ "rcg-c": container }, sizeClassNames, className);

    const getStyle = () => ({
        gap: container && adjustedGap && getGap(adjustedGap),
    });

    const renderChildren = () => {
        let currentColumns = { xs: 0, sm: 0, md: 0, lg: 0, xl: 0 };
        const childrenToRender = Array.isArray(children)
            ? children
            : [children];

        return childrenToRender.map((child, index) => {
            let renderedChild = child;

            if (child?.type?.name === Grid.name && child?.props?.item) {
                const renderedChildOffset = sizes.reduce((acc, size) => {
                    const completedClear = addMissingSizes(
                        "clear",
                        child.props.clear,
                        defaultClear,
                        useShorthandSyntax
                    );
                    const completedHide = addMissingSizes(
                        "hide",
                        child.props.hide,
                        defaultHide,
                        useShorthandSyntax
                    );
                    const completedOffset = addMissingSizes(
                        "offset",
                        child.props.offset,
                        defaultOffset,
                        useShorthandSyntax
                    );
                    const completedWidth = addMissingSizes(
                        "width",
                        child.props.width,
                        defaultWidth,
                        useShorthandSyntax
                    );
                    const { adjustedColumn, adjustedOffset } =
                        getAdjustedLayoutProps(
                            size,
                            completedOffset,
                            completedWidth,
                            completedHide,
                            completedClear,
                            currentColumns
                        );

                    currentColumns[size] = adjustedColumn;
                    acc[size] = adjustedOffset;

                    return acc;
                }, {});
                const renderedChildProps = {
                    ...child.props,
                    key: index,
                    offset: renderedChildOffset,
                };

                renderedChild = <Grid {...renderedChildProps} />;
            }

            return renderedChild;
        });
    };

    const renderWithThemeProvider = () => (
        <ThemeContext.Provider value={getThemeValues()}>
            {renderGrid()}
        </ThemeContext.Provider>
    );

    const renderGrid = () => (
        <div className={getClass()} style={getStyle()}>
            {container ? renderChildren() : children}
        </div>
    );

    return shouldUsePropGap ? renderWithThemeProvider() : renderGrid();
};

Grid.propTypes = {
    children: PropTypes.node,
    clear: PropTypes.shape({
        xs: PropTypes.bool,
        sm: PropTypes.bool,
        md: PropTypes.bool,
        lg: PropTypes.bool,
        xl: PropTypes.bool,
    }),
    className: PropTypes.string,
    container: PropTypes.bool,
    gap: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    hide: PropTypes.shape({
        xs: PropTypes.bool,
        sm: PropTypes.bool,
        md: PropTypes.bool,
        lg: PropTypes.bool,
        xl: PropTypes.bool,
    }),
    item: PropTypes.bool,
    offset: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.string,
        PropTypes.shape({
            xs: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
            sm: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
            md: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
            lg: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
            xl: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
        }),
    ]),
    width: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.string,
        PropTypes.shape({
            xs: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
            sm: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
            md: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
            lg: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
            xl: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
        }),
    ]),
};

export default Grid;
