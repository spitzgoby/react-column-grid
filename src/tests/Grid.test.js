const mockUseStyles = jest.fn();
jest.mock('react-jss', () => ({ createUseStyles: () => mockUseStyles }));

import { shallow } from 'enzyme';
import Grid from '../Grid';
import React from 'react';

describe('<Grid />', () => {
    beforeEach(() => {
        mockUseStyles.mockImplementation(() => ({}));
    });

    afterEach(() => {
        mockUseStyles.mockReset();
    });

    describe('when styling component', () => {
        it('should calculate props for all sizes when creating styles', () => {
            const hiddenProp = { xs: true, md: false };
            const offsetProp = { xs: '0', md: '2', lg: '3' };
            const widthProp = { xs: '6', md: '4', lg: '3' };

            shallow(<Grid 
                container 
                gap='2em'
                hidden={hiddenProp} 
                offset={offsetProp}
                width={widthProp}
            />);

            expect(mockUseStyles).toHaveBeenCalledWith({ 
                container: true, 
                gap: '2em',
                hidden: { xs: true, sm: true, md: false, lg: false, xl: false},
                offset: { xs: '0', sm: '0', md: '2', lg: '3', xl: '3' },
                width: { xs: '6', sm: '6', md: '4', lg: '3', xl: '3' }
            });
        });

        it('should use supplied .grid class', () => {
            mockUseStyles.mockImplementation(() => ({ grid: 'test-grid-class' }));

            const grid = shallow(<Grid item hidden={{ xs: true }}></Grid>);

            expect(grid.hasClass('test-grid-class')).toBe(true);
        });

        it('should use supplied .container class if it is a container', () => {
            mockUseStyles.mockImplementation(() => ({ container: 'test-container-class' }));

            const grid = shallow(<Grid container hidden={{ xs: true }}></Grid>);

            expect(grid.hasClass('test-container-class')).toBe(true);
        });

        it('should use supplied .item class if it is an item', () => {
            mockUseStyles.mockImplementation(() => ({ item: 'test-item-class' }));

            const grid = shallow(<Grid item hidden={{ xs: true }}></Grid>);

            expect(grid.hasClass('test-item-class')).toBe(true);
        });
    });

    describe('when the grid is a container', () => {
        it('should create a grid container', () => {
            const grid = shallow(<Grid container />);

            expect(grid).toMatchSnapshot();
        });

        it('should be able to render a single child component', () => {
            const renderGrid = () => {
                shallow(<Grid container item><div></div></Grid>);
            }

            expect(renderGrid).not.toThrow();
        });

        it('should increase the offset of child grid items in each size class', () => {
            const grid = shallow((
                <Grid container>
                    <Grid item width={{xs: 12, md: 3}} offset={{xs: 0, md: 3}} />
                    <Grid item width={{xs: 12, md: 3}} offset={{xs: 0, md: 3}} />
                </Grid>
            ));
            const expectedOffsets = [{
                xs: 0, sm: 0, md: 3, lg: 3, xl: 3
            }, {
                xs: 0, sm: 0, md: 9, lg: 9, xl: 9
            }];

            expect(grid.children().forEach((item, index) => {
                const element = item.getElement();

                Object.keys(element.props.offset).forEach(size => {
                    expect(element.props.offset[size]).toEqual(expectedOffsets[index][size]);
                });
            }));
        });

        it('should move elements onto the next row when the columns exceed row width', () => {
            const grid = shallow((
                <Grid container>
                    <Grid item width={{xs: 6, md: 3}} offset={{xs: 0, md: 3}} />
                    <Grid item width={{xs: 6, md: 3}} offset={{xs: 2, md: 3}} />
                    <Grid item width={{xs: 12, md: 3}} offset={{xs: 0, md: 3}} />
                </Grid>
            ));
            const expectedOffsets = [{
                xs: 0, sm: 0, md: 3, lg: 3, xl: 3
            }, {
                xs: 2, sm: 2, md: 9, lg: 9, xl: 9
            }, {
                xs: 0, sm: 0, md: 3, lg: 3, xl: 3
            }];

            expect(grid.children().forEach((item, index) => {
                const element = item.getElement();

                Object.keys(element.props.offset).forEach(size => {
                    expect(element.props.offset[size]).toEqual(expectedOffsets[index][size]);
                });
            }));
        });
    });
});