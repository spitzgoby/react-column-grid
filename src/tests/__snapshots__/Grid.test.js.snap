// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Grid /> when the grid is a container should create a grid container 1`] = `
<div
  className="swa-react-grid swa-react-grid--container"
/>
`;
